### PR TITLE
Move instrumentation to renderers/shared. Closes #6797

### LINKED ISSUE #6797: Move instrumentation from isomorphic to renderers/shared
Per @sebmarkbage, we want `isomorphic` folder to only contain the code necessary to author React components. It should not have dependencies on `ReactInstrumentation`, `ReactDebugTool`, and `devtools/*` friends, which themselves should be moved to `renderers/shared/`.

The only caller we to `ReactInstrumentation` we have right now in that folder is [`onSetState()`](https://github.com/facebook/react/blob/8ea1cf4ee0f5315c4190a9e67e15f0f7404cb0cf/src/isomorphic/modern/class/ReactComponent.js#L70) inside `ReactComponent`. We need to move it somewhere inside the renderer (e.g. in the update queue implementation).

Labeling it as a `good first bug` because this mostly involves moving files. The goal is that files in `isomorphic` folder don’t contain anything that references `ReactInstrumentation`.

This is up for grabs but I’ll do it myself if no one takes it.

### PR DESCRIPTION
**Pull request by iamdustan**
From the initial report I also moved ReactPerf out, which maybe should be moved back? Also you may want me to move some files around, yet. The following is the git log version of the moves:

```bash
 rename src/{isomorphic => shared}/ReactPerf.js (100%)
 rename src/{isomorphic => shared}/__tests__/ReactPerf-test.js (100%)
 rename src/{isomorphic => shared}/devtools/ReactComponentTreeDevtool.js (100%)
 rename src/{isomorphic => shared}/devtools/ReactHostOperationHistoryDevtool.js (100%)
 rename src/{isomorphic => shared}/devtools/ReactInvalidSetStateWarningDevTool.js (100%)
 rename src/{isomorphic => shared}/devtools/__tests__/ReactComponentTreeDevtool-test.js (100%)
 rename src/{isomorphic => shared}/devtools/__tests__/ReactComponentTreeDevtool-test.native.js (100%)
 rename src/{isomorphic => shared}/devtools/__tests__/ReactHostOperationHistoryDevtool-test.js (100%)
 rename src/{isomorphic => shared/utils}/ReactDebugTool.js (100%)
 rename src/{isomorphic => shared/utils}/ReactInstrumentation.js (100%)
 rename src/{isomorphic => shared/utils}/__tests__/ReactDebugTool-test.js (100%)
```

But it may be easier to see the added files in this treeview (with only added files kept):

```
src/shared
├── ReactPerf.js
├── __tests__
│   └── ReactPerf-test.js
├── devtools
│   ├── ReactComponentTreeDevtool.js
│   ├── ReactHostOperationHistoryDevtool.js
│   ├── ReactInvalidSetStateWarningDevTool.js
│   └── __tests__
│       ├── ReactComponentTreeDevtool-test.js
│       ├── ReactComponentTreeDevtool-test.native.js
│       └── ReactHostOperationHistoryDevtool-test.js
└── utils
    ├── ReactDebugTool.js
    ├── ReactInstrumentation.js
    └── __tests__
          └── ReactDebugTool-test.js

```

### CODE REVIEW

File: src/isomorphic/modern/class/ReactComponent.js [hunk 1]
```diff
@@ -66,14 +65,6 @@ ReactComponent.prototype.setState = function(partialState, callback) {
...
   );
-  if (__DEV__) {
-    ReactInstrumentation.debugTool.onSetState();
-    warning(
-      partialState != null,
-      'setState(...): You passed an undefined or null state object; ' +
-      'instead, use forceUpdate().'
-    );
-  }
   this.updater.enqueueSetState(this, partialState);
```
Comment (iamdustan) [inline src/isomorphic/modern/class/ReactComponent.js]: `this.updater` falls back to the `ReactNoopUpdateQueue`which is also in `isomorphic/modern/class/`. Is it okay to not call `ReactInstrumentation.debugTool.onSetState()` in that scenario?

Comment (gaearon) [inline src/isomorphic/modern/class/ReactComponent.js]: >Is it okay to not call ReactInstrumentation.debugTool.onSetState() in that scenario?

Yea, I think it’s fine.

**Comment by gaearon** [comment]
Hmm. Doesn’t this move to `src/shared` rather than `src/renderers/shared`?

**Comment by iamdustan** [comment]
Doh. Updated.

**Git view**

```
 rename src/{isomorphic => renderers/shared}/ReactDebugTool.js (100%)
 rename src/{isomorphic => renderers/shared}/ReactInstrumentation.js (100%)
 rename src/{isomorphic => renderers/shared}/ReactPerf.js (100%)
 rename src/{isomorphic => renderers/shared}/__tests__/ReactDebugTool-test.js (100%)
 rename src/{isomorphic => renderers/shared}/__tests__/ReactPerf-test.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/ReactComponentTreeDevtool.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/ReactHostOperationHistoryDevtool.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/ReactInvalidSetStateWarningDevTool.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/__tests__/ReactComponentTreeDevtool-test.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/__tests__/ReactComponentTreeDevtool-test.native.js (100%)
 rename src/{isomorphic => renderers/shared}/devtools/__tests__/ReactHostOperationHistoryDevtool-test.js (100%)
```

**(Minimal) Tree view**

```
src/renderers/shared
├── ReactDebugTool.js
├── ReactInstrumentation.js
├── ReactPerf.js
├── __tests__
│   ├── ReactDebugTool-test.js
│   └── ReactPerf-test.js
└── devtools
    ├── ReactComponentTreeDevtool.js
    ├── ReactHostOperationHistoryDevtool.js
    ├── ReactInvalidSetStateWarningDevTool.js
    └── __tests__
        ├── ReactComponentTreeDevtool-test.js
        ├── ReactComponentTreeDevtool-test.native.js
        └── ReactHostOperationHistoryDevtool-test.js
```

**Comment by facebook-github-bot-6** [comment]
@iamdustan updated the pull request.

**Pull request merged**